File: LeetCode/528.Random_Pick_with_Weight/528.Random_Pick_with_Weight.swift

class Solution {
    var prefix: [Int]
    let total: Int

    init(_ w: [Int]) {
        var run = 0
        prefix = w.map { e in
            run += e
            return run
        }
        total = run
    }

    func pickIndex() -> Int {
        let t = Int.random(in: 1 ... total)
        var l = 0, r = prefix.count
        while l < r {
            let m = l + (r - l) >> 1
            if prefix[m] < t {
                l = m + 1
            } else {
                r = m
            }
        }
        return l
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * let obj = Solution(w)
 * let ret_1: Int = obj.pickIndex()
 */

File: LeetCode/528.Random_Pick_with_Weight/528.Random_Pick_with_Weight.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from typing import List
import random
import bisect


class Solution:

    def __init__(self, w: List[int]):
        self.prefix = []
        self.prefix.append(w[0])
        for i in range(1, len(w)):
            self.prefix.append(self.prefix[-1] + w[i])

    def pickIndex(self) -> int:
        val = random.randint(1, self.prefix[-1])
        idx = bisect.bisect_left(self.prefix, val)
        return idx
